File: packages/platools-py/platools/core/decorator.py
"""`@platools.tool()` decorator implementation.

The decorator:
  1. Inspects the wrapped function
  2. Builds input + output JSON Schemas via `platools.core.schema`
  3. Creates a `ToolDef` carrying name, description, schemas, auth, roles,
     rate_limit, timeout, annotations
  4. Registers the tool in the `Platools` instance's registry
  5. Returns the original function unchanged — decorated functions remain
     directly callable in user code

Supports both sync and async functions.
"""

from __future__ import annotations

import inspect
from collections.abc import Callable
from typing import Any, TypeVar

from platools.core.registry import ToolRegistry
from platools.core.schema import auto_description, build_input_schema, build_output_schema
from platools.types import AuthLevel, ToolDef

F = TypeVar("F", bound=Callable[..., Any])


_VALID_AUTH: frozenset[str] = frozenset(("none", "user", "admin"))

# CTX.5: parameter names the decorator recognizes as "inject the
# `PlatosContext` here". ``ctx`` is the canonical name; ``platos_ctx``
# is offered as a namespaced alternative for handlers that already use
# ``ctx`` for something else.
_CTX_PARAM_NAMES: frozenset[str] = frozenset(("ctx", "platos_ctx"))


def _detect_ctx_param(func: Callable[..., Any]) -> str | None:
    """Return the parameter name to inject the :class:`PlatosContext` into.

    CTX.5 opt-in: if the function signature has a parameter named
    ``ctx`` (or ``platos_ctx``), the transport layer injects the
    per-call context under that kwarg; the tool-call params arriving
    over the wire never carry ``ctx`` themselves. Returns ``None`` when
    the handler doesn't opt in, in which case the transport dispatches
    via kwargs only (historical behavior preserved).
    """
    import inspect as _inspect

    try:
        sig = _inspect.signature(func)
    except (TypeError, ValueError):
        return None
    for name in sig.parameters:
        if name in _CTX_PARAM_NAMES:
            return name
    return None


def make_tool_decorator(registry: ToolRegistry) -> Callable[..., Callable[[F], F]]:
    """Return a decorator factory bound to the given registry.

    `Platools.tool` is assigned from this so each `Platools()` instance
    has its own registry while sharing the same decorator logic.
    """

    def tool(
        *,
        name: str | None = None,
        description: str | None = None,
        auth: AuthLevel = "none",
        roles: list[str] | tuple[str, ...] | None = None,
        rate_limit: str | None = None,
        timeout: int | None = None,
        annotations: dict[str, Any] | None = None,
    ) -> Callable[[F], F]:
        if auth not in _VALID_AUTH:
            raise ValueError(f"auth must be one of {sorted(_VALID_AUTH)}, got {auth!r}")
        if timeout is not None and timeout <= 0:
            raise ValueError(f"timeout must be positive, got {timeout}")

        def decorator(func: F) -> F:
            tool_name = name or func.__name__
            tool_description = description or auto_description(func)
            input_schema = build_input_schema(func)
            output_schema = build_output_schema(func)
            tool_def = ToolDef(
                name=tool_name,
                description=tool_description,
                func=func,
                input_schema=input_schema,
                output_schema=output_schema,
                auth=auth,
                roles=tuple(roles or ()),
                rate_limit=rate_limit,
                timeout=timeout,
                annotations=dict(annotations or {}),
                is_async=inspect.iscoroutinefunction(func),
                ctx_param_name=_detect_ctx_param(func),
            )
            registry.register(tool_def)
            return func

        return decorator

    return tool

File: packages/platools-js/src/core/decorator.ts
/**
 * `platools.tool()` factory — the TypeScript analogue of the Python
 * `@platools.tool()` decorator.
 *
 * TC39 decorators are still shifting under our feet, so the TS SDK
 * uses a plain factory function (same pattern tRPC, Hono, and Zod
 * itself use). Consumers write:
 *
 *     export const processRefund = platools.tool(
 *       {
 *         name: "process_refund",
 *         description: "Process a refund for an order",
 *         input: z.object({ orderId: z.string(), reason: z.string() }),
 *         output: RefundResult,
 *         auth: "user",
 *         roles: ["support", "admin"],
 *       },
 *       async ({ orderId, reason }, ctx) => refundService.process(orderId, reason),
 *     );
 *
 * The factory:
 *
 *   1. Validates the `ToolOptions` (auth level, timeout, roles).
 *   2. Builds input + output JSON schemas via `zod-to-json-schema`.
 *   3. Registers a `ToolDef` in the owning `Platools` instance's
 *      registry.
 *   4. Returns the handler unchanged — decorated tools remain
 *      directly callable in user code, exactly like the Python SDK.
 *
 * Sync and async handlers are both supported; the transport
 * `await`s the handler's return value unconditionally.
 */

import type { z } from "zod";

import { buildPlatosContext, type PlatosContext } from "../context.js";
import { buildSchemas } from "./schema.js";
import { ToolRegistry } from "./registry.js";
import type {
  AuthLevel,
  ToolDef,
  ToolHandler,
  ToolOptions,
} from "../types.js";

const VALID_AUTH: ReadonlySet<AuthLevel> = new Set<AuthLevel>(["none", "user", "admin"]);

/**
 * Build a `tool()` factory bound to `registry`. Each `Platools`
 * instance has its own registry, so every instance's `.tool()` is a
 * closure over its own registry reference — two `Platools` objects
 * never leak registrations to each other.
 */
export function makeToolFactory(
  registry: ToolRegistry,
): <Input extends z.ZodTypeAny, Output extends z.ZodTypeAny>(
  options: ToolOptions<Input, Output>,
  handler: ToolHandler<z.infer<Input>, z.infer<Output>>,
) => ToolHandler<z.infer<Input>, z.infer<Output>> {
  return function tool<Input extends z.ZodTypeAny, Output extends z.ZodTypeAny>(
    options: ToolOptions<Input, Output>,
    handler: ToolHandler<z.infer<Input>, z.infer<Output>>,
  ): ToolHandler<z.infer<Input>, z.infer<Output>> {
    validateOptions(options);

    const { inputSchema, outputSchema } = buildSchemas(options);

    // Wrap the handler in a common `unknown`-typed shim so the
    // registry stays homogeneous. The original, properly-typed
    // handler is returned to the caller untouched.
    const boxed: ToolHandler<unknown, unknown> = (params, ctx) =>
      handler(params as z.infer<Input>, ctx);

    const toolDef: ToolDef = {
      name: options.name,
      description: (options.description ?? "").trim(),
      handler: boxed,
      inputZodSchema: options.input,
      outputZodSchema: options.output ?? null,
      inputSchema,
      outputSchema,
      auth: options.auth ?? "none",
      roles: Object.freeze([...(options.roles ?? [])]),
      rateLimit: options.rateLimit ?? null,
      timeoutMs: options.timeoutMs ?? null,
      annotations: Object.freeze({ ...(options.annotations ?? {}) }),
    };
    registry.register(toolDef);

    // Return the consumer's handler unchanged — it remains directly
    // callable in user code. This matches the Python decorator's
    // "returns original function" behavior so unit tests can
    // invoke the tool without going through the registry.
    return handler;
  };
}

function validateOptions<Input extends z.ZodTypeAny, Output extends z.ZodTypeAny>(
  options: ToolOptions<Input, Output>,
): void {
  if (!options.name || options.name.trim() === "") {
    throw new Error("platools.tool: `name` is required");
  }
  const auth = options.auth ?? "none";
  if (!VALID_AUTH.has(auth)) {
    throw new Error(
      `platools.tool: auth must be one of ${[...VALID_AUTH].sort().join(", ")}, got ${auth}`,
    );
  }
  if (options.timeoutMs !== undefined && options.timeoutMs <= 0) {
    throw new Error(
      `platools.tool: timeoutMs must be positive, got ${options.timeoutMs}`,
    );
  }
  if (!options.input) {
    throw new Error("platools.tool: `input` (Zod schema) is required");
  }
}

/**
 * Build a no-op `PlatosContext`. Used by tests and by the local test
 * runner which doesn't go through the WebSocket transport.
 *
 * CTX.5: the returned shape now matches the full handler-ctx
 * contract (`callId`, `context`, `entityIds?`, `raw`) so local tests
 * and the platform-dispatched path see an identical type.
 */
export function makeLocalContext(callId = "local"): PlatosContext {
  return buildPlatosContext(callId, {});
}
